pyogrio/_io.pyx

```diff
@@ -1126,6 +1126,14 @@ def ogr_read(
         if sql is None:
             # layer defaults to index 0
             if layer is None:
+                layers = ogr_list_layers_dataset(ogr_dataset)
```

Comment:
This block could probably be consolidated into a function:

```
cdef get_default_layer(OGRDataSourceH ogr_dataset):
    layers = ogr_list_layers_dataset(ogr_dataset)
    first_layer_name = layers[0][0]

    if len(layers) > 1:
        dataset_name = os.path.basename(OGR_DS_GetName(ogr_dataset).decode('UTF-8'))

        other_layer_names = ', '.join([f"'{l}'" for l in layers[1:, 0]])
        warnings.warn(
            f"More than one layer found in '{dataset_name}': '{first_layer_name}' (default), {other_layer_names}.  "
            "Specify layer parameter to avoid this warning.",
            stacklevel=2,
        )

    return first_layer_name
```
(not tested against all the various VSI filenames, but worked properly for the `test_fgdb_vsi` fixture)


You'll need to add the following to `_ogr.pxd` in the `cdef extern from "ogr_api.h":` block:

```
const char*     OGR_DS_GetName(OGRDataSourceH)
```


This returns a somewhat more succinct warning:

```
UserWarning: More than one layer found in 'sample.osm.pbf': 'points' (default), 'lines', 'multilinestrings', 'multipolygons', 'other_relations'.  Specify layer parameter to avoid this warning.
```

Reply:
That's indeed a lot cleaner. Done.